### PR TITLE
[SPARK-9832] [SQL] add a thread-safe lookup for BytesToBytseMap

### PR DESCRIPTION
**Pull request by davies**
This patch add a thread-safe lookup for BytesToBytseMap, and use that in broadcasted HashedRelation.

### CODE REVIEW

**Comment by davies** [comment]
cc @JoshRosen

File: core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java [hunk 1]
```diff
@@ -328,6 +327,20 @@ public Location lookup(
       Object keyBaseObject,
       long keyBaseOffset,
       int keyRowLengthBytes) {
+    safeLookup(keyBaseObject, keyBaseOffset, keyRowLengthBytes, loc);
+    return loc;
+  }
+
+  /**
+   * Looks up a key, and saves the result in provided `loc` and returns it.
```
Comment (JoshRosen) [inline core/src/main/java/org/apache/spark/unsafe/map/BytesToBytesMap.java]: This doesn't return the `loc`; it's void.

**Comment by JoshRosen** [comment]
LGTM.

**Comment by rxin** [comment]
I'm going to merge thsi.

**Pull request closed without merging**